### PR TITLE
arch/risc-v/riscv_misaligned: Implement float load/store support

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
 Implement misaligned float load/store
## Impact
risc-v with FPU support
## Testing
K210

### CODE REVIEW

File: arch/risc-v/src/common/riscv_misaligned.c [hunk 1]
```diff
@@ -443,9 +518,32 @@ static bool decode_insn(uintptr_t *regs, riscv_insn_ctx_t *ctx)
...
+
+        imm = sext(insn.s.imm2 | insn.s.imm1 << 5, 12);
+
+        ctx->dest = (uint8_t *)regs[insn.s.rs1] + imm;
+        ctx->src = (uint8_t *)&regs[REG_F0 + insn.s.rs2];
+
+        /* Is instruction fsw or fsd ? */
+
+        ctx->len = insn.s.funct3 == 0x2 ? 4 : 8;
+
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_misaligned.c:544]:
```suggestion
```

**Review by pkarashchenko — APPROVED** [review]

File: arch/risc-v/src/common/riscv_misaligned.c [hunk 2]
```diff
@@ -443,9 +518,31 @@ static bool decode_insn(uintptr_t *regs, riscv_insn_ctx_t *ctx)
 
       case INSN_FLW:
       case INSN_FLD:
+
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_misaligned.c:521]: wrap with `#ifdef CONFIG_ARCH_FPU` is needed.
Also please remove TODO from `/* TODO: Handle float load / store instruction */`

Comment (no1wudi) [reply arch/risc-v/src/common/riscv_misaligned.c:520]: Yes, gurad added.

File: arch/risc-v/src/common/riscv_misaligned.c [hunk 3]
```diff
@@ -329,17 +329,92 @@ static bool decode_insn_compressed(uintptr_t *regs, riscv_insn_ctx_t *ctx)
...
+        imm = insn.lwsp.imm42 << 2 | insn.lwsp.imm5 << 5 |
+              insn.lwsp.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_F0 + insn.lwsp.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 4;
+        break;
+
       case INSN_C_FSW:
+
+        /* fsw share the same encoding layout with swsp */
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_misaligned.c:357]: swsp->sw?

File: arch/risc-v/src/common/riscv_misaligned.c [hunk 4]
```diff
@@ -329,17 +329,92 @@ static bool decode_insn_compressed(uintptr_t *regs, riscv_insn_ctx_t *ctx)
...
+        imm = insn.sw.imm2 << 2 | insn.sw.imm53 << 3 | insn.sw.imm6 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sw.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_F8 + insn.sd.rs2];
+        ctx->len  = 4;
+        break;
+
       case INSN_C_FSWSP:
-#  endif
+
+        /* fswsp share the same encoding layout with fswsp */
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_misaligned.c:367]: fswsp->swsp

File: arch/risc-v/src/common/riscv_misaligned.c [hunk 5]
```diff
@@ -439,13 +514,35 @@ static bool decode_insn(uintptr_t *regs, riscv_insn_ctx_t *ctx)
 
         break;
 
-      /* TODO: Handle float load / store instruction */
-
+#ifdef CONFIG_ARCH_FPU
       case INSN_FLW:
       case INSN_FLD:
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_misaligned.c:519]: Do you have plans to support `INSN_FLQ` / `INSN_FSQ`

**Review by pkarashchenko — APPROVED** [review]
LGTM after fixing issues in comments.

File: arch/risc-v/src/common/riscv_misaligned.c (hunk 5 above)
Comment (no1wudi) [reply arch/risc-v/src/common/riscv_misaligned.c:519]: Yes if any device support it

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**